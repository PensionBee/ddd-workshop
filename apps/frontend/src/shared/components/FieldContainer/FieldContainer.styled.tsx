import styled from "styled-components";
import { Box } from "@shared/components/Box";
import { StyledErrorMessage } from "./ErrorMessage/ErrorMessage.styled";

export const StyledFieldContainer = styled(Box)`
  margin-bottom: 36px;
  &:focus-within {
    ${StyledErrorMessage} {
      display: none;
    }
  }
`;
