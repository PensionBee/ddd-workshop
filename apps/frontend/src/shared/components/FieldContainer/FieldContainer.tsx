import { ErrorMessage } from "./ErrorMessage/ErrorMessage";
import type { LabelProps } from "./Label/Label";
import { Label } from "./Label/Label";
import type { HintTextProps } from "./HintText/HintText";
import { HintText } from "./HintText/HintText";
import { Box } from "@shared/components/Box";
import { StyledFieldContainer } from "./FieldContainer.styled";

type BaseFieldContainerProps = {
  /** The field(s) to wrap */
  children: React.ReactNode;
  /** The error message to display for the input */
  error?: string;
};

export type FieldContainerProps = BaseFieldContainerProps &
  LabelProps &
  HintTextProps;

export const FieldContainer: React.FC<FieldContainerProps> = ({
  children,
  error,
  label,
  labelHref,
  labelLinkText,
  labelTooltip,
  labelHtmlFor,
  hintText,
  ...props
}) => {
  const labelProps = {
    label,
    labelHref,
    labelLinkText,
    labelTooltip,
    labelHtmlFor,
  } satisfies LabelProps;

  const hintTextProps = {
    hintText,
  } satisfies HintTextProps;

  return (
    <StyledFieldContainer {...props}>
      <Label {...labelProps} />
      <Box d="flex" w="100%" rowGap={16} colGap={16} justify="stretch">
        {children}
      </Box>
      <ErrorMessage>{error}</ErrorMessage>
      <HintText {...hintTextProps} />
    </StyledFieldContainer>
  );
};
