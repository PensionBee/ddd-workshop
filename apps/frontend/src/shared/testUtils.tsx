import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
