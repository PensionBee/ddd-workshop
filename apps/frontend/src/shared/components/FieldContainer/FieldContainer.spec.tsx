import { render, vi } from "@shared/testUtils";
import { FieldContainer } from "./FieldContainer";

const INPUT_CONTENT = "Input content";
const LABEL = "Label";
const HINT_TEXT = "Hint text";

describe("FieldContainer", () => {
  describe("Default", () => {
    it("should render correctly", () => {
      const { getByText, asFragment } = render(
        <FieldContainer>
          <div>{INPUT_CONTENT}</div>
        </FieldContainer>
      );

      expect(getByText(INPUT_CONTENT)).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("With label", () => {
    it("should render correctly", () => {
      const { getByText, asFragment } = render(
        <FieldContainer label={LABEL}>
          <div>{INPUT_CONTENT}</div>
        </FieldContainer>
      );

      expect(getByText(INPUT_CONTENT)).toBeVisible();
      expect(getByText(LABEL)).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("With hint text", () => {
    it("should render correctly", () => {
      const { getByText, asFragment } = render(
        <FieldContainer hintText={HINT_TEXT}>
          <div>{INPUT_CONTENT}</div>
        </FieldContainer>
      );

      expect(getByText(HINT_TEXT)).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("With error message", () => {
    it("should render correctly", () => {
      const { getByText, asFragment } = render(
        <FieldContainer label="Label" error="Error message">
          <div>{INPUT_CONTENT}</div>
        </FieldContainer>
      );

      expect(getByText(INPUT_CONTENT)).toBeVisible();
      expect(getByText("Error message")).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
