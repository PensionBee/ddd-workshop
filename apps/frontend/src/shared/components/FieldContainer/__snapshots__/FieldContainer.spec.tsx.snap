// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FieldContainer > Default > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gsnScm hcHchc sc-iBkiHk jlUpdZ"
  >
    <div
      class="sc-gsnScm gHKBVn"
    >
      <div>
        Input content
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FieldContainer > With error message > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gsnScm hcHchc sc-iBkiHk jlUpdZ"
  >
    <label
      class="sc-gsnScm dmxLJN"
    >
      <span
        class="sc-bczSft iwUwuA"
        data-rc="BodySmallB"
      >
        Label
      </span>
    </label>
    <div
      class="sc-gsnScm gHKBVn"
    >
      <div>
        Input content
      </div>
    </div>
    <div
      class="sc-gsnScm cUcvaQ sc-dkzCjb kZxHOV"
    >
      <span
        class="sc-bczSft imQpyU"
        data-rc="DisclaimerR"
      >
        Error message
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FieldContainer > With hint text > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gsnScm hcHchc sc-iBkiHk jlUpdZ"
  >
    <div
      class="sc-gsnScm gHKBVn"
    >
      <div>
        Input content
      </div>
    </div>
    <span
      class="sc-bczSft jqHUNi"
      data-rc="DisclaimerR"
    >
      Hint text
    </span>
  </div>
</DocumentFragment>
`;

exports[`FieldContainer > With label > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gsnScm hcHchc sc-iBkiHk jlUpdZ"
  >
    <label
      class="sc-gsnScm dmxLJN"
    >
      <span
        class="sc-bczSft iwUwuA"
        data-rc="BodySmallB"
      >
        Label
      </span>
    </label>
    <div
      class="sc-gsnScm gHKBVn"
    >
      <div>
        Input content
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FieldContainer Default should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR bANfYu sc-iBYQkv dNyaQo"
  >
    <div
      class="sc-gswNZR joFDJF"
    >
      <div>
        Input content
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FieldContainer With error message should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR bANfYu sc-iBYQkv dNyaQo"
  >
    <label
      class="sc-gswNZR gbpcZv"
    >
      <span
        class="sc-bcXHqe fLXyHo"
        data-rc="BodySmallB"
      >
        Label
      </span>
    </label>
    <div
      class="sc-gswNZR joFDJF"
    >
      <div>
        Input content
      </div>
    </div>
    <div
      class="sc-gswNZR bwgTNC sc-dkrFOg kvDetP"
    >
      <span
        class="sc-bcXHqe hVJkQo"
        data-rc="DisclaimerR"
      >
        Error message
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FieldContainer With hint text should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR bANfYu sc-iBYQkv dNyaQo"
  >
    <div
      class="sc-gswNZR joFDJF"
    >
      <div>
        Input content
      </div>
    </div>
    <span
      class="sc-bcXHqe TOXwu"
      data-rc="DisclaimerR"
    >
      Hint text
    </span>
  </div>
</DocumentFragment>
`;

exports[`FieldContainer With label should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR bANfYu sc-iBYQkv dNyaQo"
  >
    <label
      class="sc-gswNZR gbpcZv"
    >
      <span
        class="sc-bcXHqe fLXyHo"
        data-rc="BodySmallB"
      >
        Label
      </span>
    </label>
    <div
      class="sc-gswNZR joFDJF"
    >
      <div>
        Input content
      </div>
    </div>
  </div>
</DocumentFragment>
`;
